File: js/src/core/security/yeria-user-token-verifier.ts
import { createVerify } from 'crypto';
import { SignatureVerificationError, ViewExpiredError } from '../../errors';
import { YeriaTokenClaims, YeriaPublicKeyResolver } from '../yeria-protocol';

/**
 * Verifies a Yeria-ISSUED USER JWT (RS256): enforces `iss='yeria'`, an optional
 * `aud=<serviceId>`, and `exp > now`. Static-only helper.
 *
 * `verifyYeriaToken(jwt, pem, aud?)` verifies against a PEM you supply;
 * `verifyYeriaTokenWithResolver(jwt, resolver, aud?)` resolves the key from the
 * token's `kid` header via a resolver (typically YeriaPublicKeys.getByKid).
 * Throws SignatureVerificationError / ViewExpiredError on any failure.
 *
 * Not: this verifies Yeria user JWTs only — it is distinct from YeriaSigner
 * (signs), YeriaEnvelopeVerifier (verifies provider view envelopes) and
 * YeriaPublicKeys (which only resolves keys).
 */
export class YeriaUserTokenVerifier {
    /** Verify a Yeria user JWT against a PEM you supply; returns the claims. */
    static verifyYeriaToken(
        jwtToken: string,
        yeriaPublicKey: string,
        expectedAudience?: string | number,
    ): YeriaTokenClaims {
        if (typeof jwtToken !== 'string' || jwtToken.length === 0) {
            throw new SignatureVerificationError('yeria', 'missing jwt');
        }
        const parts = jwtToken.split('.');
        if (parts.length !== 3 || !parts[0] || !parts[1] || !parts[2]) {
            throw new SignatureVerificationError('yeria', 'malformed jwt');
        }
        const headerB64 = parts[0];
        const payloadB64 = parts[1];
        const signatureB64 = parts[2];

        let header: Record<string, unknown>;
        let claims: YeriaTokenClaims;
        try {
            header = JSON.parse(base64UrlDecodeToString(headerB64));
            claims = JSON.parse(base64UrlDecodeToString(payloadB64)) as YeriaTokenClaims;
        } catch (_e) {
            throw new SignatureVerificationError('yeria', 'invalid base64 / json');
        }

        if (header['alg'] !== 'RS256') {
            throw new SignatureVerificationError(
                'yeria',
                `unsupported alg: ${String(header['alg'])}`,
            );
        }

        const signingInput = `${headerB64}.${payloadB64}`;
        const signatureBytes = base64UrlDecodeToBuffer(signatureB64);
        const verifier = createVerify('RSA-SHA256');
        verifier.update(signingInput);
        verifier.end();
        const sigOk = verifier.verify(yeriaPublicKey, signatureBytes);
        if (!sigOk) {
            throw new SignatureVerificationError('yeria', 'signature mismatch');
        }

        if (claims.iss !== 'yeria') {
            throw new SignatureVerificationError(
                'yeria',
                `unexpected issuer: ${String(claims.iss)}`,
            );
        }
        if (expectedAudience !== undefined && claims.aud !== String(expectedAudience)) {
            throw new SignatureVerificationError(
                'yeria',
                `audience mismatch: token aud=${claims.aud}, expected=${String(expectedAudience)}`,
            );
        }

        const nowSec = Math.floor(Date.now() / 1000);
        if (typeof claims.exp !== 'number' || claims.exp <= nowSec) {
            throw new ViewExpiredError('user-token', (nowSec - (claims.exp || 0)) * 1000, 0);
        }

        if (typeof header['kid'] === 'string') {
            claims.kid = header['kid'] as string;
        }
        return claims;
    }

    /** Verify a Yeria user JWT, resolving the key from its `kid` header via `resolver`. */
    static async verifyYeriaTokenWithResolver(
        jwtToken: string,
        resolver: YeriaPublicKeyResolver,
        expectedAudience?: string | number,
    ): Promise<YeriaTokenClaims> {
        if (typeof jwtToken !== 'string' || jwtToken.length === 0) {
            throw new SignatureVerificationError('yeria', 'missing jwt');
        }
        const parts = jwtToken.split('.');
        if (parts.length !== 3 || !parts[0] || !parts[1] || !parts[2]) {
            throw new SignatureVerificationError('yeria', 'malformed jwt');
        }

        let header: Record<string, unknown>;
        try {
            header = JSON.parse(base64UrlDecodeToString(parts[0]));
        } catch (_e) {
            throw new SignatureVerificationError('yeria', 'invalid header');
        }

        const kid = header['kid'];
        if (typeof kid !== 'string' || kid.length === 0) {
            throw new SignatureVerificationError('yeria', 'jwt header missing kid');
        }

        const pem = await resolver(kid);
        if (!pem) {
            throw new SignatureVerificationError(
                'yeria',
                `no trusted key for kid=${kid}`,
            );
        }

        return YeriaUserTokenVerifier.verifyYeriaToken(jwtToken, pem, expectedAudience);
    }
}

function base64UrlDecodeToBuffer(input: string): Buffer {
    const padded = input.replace(/-/g, '+').replace(/_/g, '/') +
        '==='.slice((input.length + 3) % 4);
    return Buffer.from(padded, 'base64');
}

function base64UrlDecodeToString(input: string): string {
    return base64UrlDecodeToBuffer(input).toString('utf8');
}

File: py/yeriasdk/core/security/yeria_user_token_verifier.py
"""YeriaUserTokenVerifier — verifies Yeria-issued RS256 user tokens.
Mirrors js/src/core/security/yeria-user-token-verifier.ts.

`verify_yeria_token(jwt, pem, expected_audience?)` is pure/local (you supply the
key). `verify_yeria_token_with_resolver(jwt, resolver, expected_audience?)`
pulls the key by the token's `kid` via a resolver (e.g. YeriaPublicKeys).
"""

import base64
import json
import time
from typing import Any, Optional

from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives.hashes import SHA256
from cryptography.hazmat.backends import default_backend

from ..yeria_protocol import YeriaTokenClaims, PublicKeyResolver
from ...errors.exceptions import SignatureVerificationError, ViewExpiredError


def _b64url_decode(data: str) -> bytes:
    padded = data + "=" * ((4 - len(data) % 4) % 4)
    return base64.urlsafe_b64decode(padded)


class YeriaUserTokenVerifier:
    """Verifies a Yeria-ISSUED USER JWT (RS256): enforces ``iss='yeria'``, an
    optional ``aud=<serviceId>``, and ``exp > now``. Static-only helper.

    ``verify_yeria_token(jwt, pem, expected_audience?)`` verifies against a PEM
    you supply; ``verify_yeria_token_with_resolver(jwt, resolver,
    expected_audience?)`` resolves the key from the token's ``kid`` header via a
    resolver (e.g. YeriaPublicKeys). Raises SignatureVerificationError /
    ViewExpiredError on any failure.

    Not: this verifies Yeria user JWTs only -- it is distinct from YeriaSigner
    (signs), YeriaEnvelopeVerifier (verifies provider view envelopes) and
    YeriaPublicKeys (which only resolves keys).
    """

    @staticmethod
    def verify_yeria_token(
        jwt_token: str,
        yeria_public_key: str,
        expected_audience: Optional[Any] = None,
    ) -> YeriaTokenClaims:
        """Verify a Yeria user JWT against a PEM you supply; returns the claims."""
        if not isinstance(jwt_token, str) or not jwt_token:
            raise SignatureVerificationError("yeria", "missing jwt")

        parts = jwt_token.split(".")
        if len(parts) != 3 or not all(parts):
            raise SignatureVerificationError("yeria", "malformed jwt")
        header_b64, payload_b64, signature_b64 = parts

        try:
            header = json.loads(_b64url_decode(header_b64).decode("utf-8"))
            claims_raw = json.loads(_b64url_decode(payload_b64).decode("utf-8"))
        except Exception:
            raise SignatureVerificationError("yeria", "invalid base64 / json")

        if header.get("alg") != "RS256":
            raise SignatureVerificationError("yeria", f"unsupported alg: {header.get('alg')!r}")

        signing_input = f"{header_b64}.{payload_b64}".encode("utf-8")
        public_key_obj = serialization.load_pem_public_key(yeria_public_key.encode(), backend=default_backend())
        try:
            public_key_obj.verify(_b64url_decode(signature_b64), signing_input, padding.PKCS1v15(), SHA256())
        except Exception:
            raise SignatureVerificationError("yeria", "signature mismatch")

        if claims_raw.get("iss") != "yeria":
            raise SignatureVerificationError("yeria", f"unexpected issuer: {claims_raw.get('iss')!r}")
        if expected_audience is not None and claims_raw.get("aud") != str(expected_audience):
            raise SignatureVerificationError(
                "yeria",
                f"audience mismatch: token aud={claims_raw.get('aud')!r}, expected={str(expected_audience)!r}",
            )

        now_sec = int(time.time())
        exp = claims_raw.get("exp")
        if not isinstance(exp, int) or exp <= now_sec:
            raise ViewExpiredError("user-token", (now_sec - (exp or 0)) * 1000, 0)

        return YeriaTokenClaims(
            sub=str(claims_raw.get("sub", "")),
            aud=str(claims_raw.get("aud", "")),
            iss=str(claims_raw.get("iss", "")),
            exp=int(exp),
            iat=int(claims_raw["iat"]) if "iat" in claims_raw else None,
            kid=str(header["kid"]) if isinstance(header.get("kid"), str) else None,
        )

    @staticmethod
    def verify_yeria_token_with_resolver(
        jwt_token: str,
        resolver: PublicKeyResolver,
        expected_audience: Optional[Any] = None,
    ) -> YeriaTokenClaims:
        """Verify a Yeria user JWT, resolving the key from its ``kid`` header via ``resolver``."""
        if not isinstance(jwt_token, str) or not jwt_token:
            raise SignatureVerificationError("yeria", "missing jwt")

        parts = jwt_token.split(".")
        if len(parts) != 3 or not all(parts):
            raise SignatureVerificationError("yeria", "malformed jwt")

        try:
            header = json.loads(_b64url_decode(parts[0]).decode("utf-8"))
        except Exception:
            raise SignatureVerificationError("yeria", "invalid header")

        kid = header.get("kid")
        if not isinstance(kid, str) or not kid:
            raise SignatureVerificationError("yeria", "jwt header missing kid")

        pem = resolver(kid)
        if not pem:
            raise SignatureVerificationError("yeria", f"no trusted key for kid={kid}")

        return YeriaUserTokenVerifier.verify_yeria_token(jwt_token, pem, expected_audience)
